Migrate EliminateCrossJoins to iterative optimizer

findepi opened the pull request:
Fixes #7172, depends on #7982 (extracted from this one)

findepi commented:
cc @rschlussel

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java:
```diff
@@ -145,14 +146,14 @@ public Rewriter(PlanNodeIdAllocator idAllocator, JoinGraph graph, List<Integer>
             this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
             this.graph = requireNonNull(graph, "graph is null");
-            this.joinOrder = requireNonNull(joinOrder, "joinOrder is null");
+            this.joinOrder = ImmutableList.copyOf(requireNonNull(joinOrder, "joinOrder is null"));
```
does not `copyOf` have its own `requireNonNull` inside?

findepi replied:
It does, but the exc msg would be less informative. I think I follow established convention here.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java:
```diff
@@ -53,8 +54,25 @@
+    /**
+     * Build a non-trivial (more than 1 element) {@link JoinGraph} containing {@code plan}.
+     */
+    public static Optional<JoinGraph> buildShallowFrom(PlanNode plan, Lookup lookup)
```
what does the `Shallow` mean here?

findepi replied:
That it builds only top-level `JoinGraph`. Compare that to the already existing `buildFrom` which builds all `JoinGraphs` from the plan tree.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java:
```diff
@@ -0,0 +1,75 @@
+                .on(this::joinAndCrossJoin)
+                .matches(
+                        project(
+                                join(JoinNode.Type.INNER,
```
static import

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java:
```diff
@@ -0,0 +1,75 @@
+                .matches(
+                        project(
+                                join(JoinNode.Type.INNER,
+                                        ImmutableList.of(aliases -> new JoinNode.EquiJoinClause(new Symbol("cySymbol"), new Symbol("bySymbol"))),
```
it would be nice to have some DSL method for equiJoinClause

findepi replied:
`join(..)` does not take a list of `EquiJoinClause`-s, it takes list of factories of `EquiJoinClause`. We have a DSL method for the latter (`com.facebook.presto.sql.planner.assertions.PlanMatchPattern#equiJoinClause`), but this case is a bit special and I cannot use that.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java:
```diff
@@ -0,0 +1,75 @@
+    ));
+
+    @Test
+    public void testEliminateCrossJoin()
```
can you please add some tests with OUTER join?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java:
```diff
@@ -200,15 +218,28 @@ private JoinGraph joinWith(JoinGraph other, List<JoinNode.EquiJoinClause> joinCl
     private static class Builder
             extends PlanVisitor<Context, JoinGraph>
     {
+        // TODO When com.facebook.presto.sql.planner.optimizations.EliminateCrossJoins is removed, remove 'shallow' flag
```
using Visitors with a hacked Lookup is acceptable to me when that Visitor is calculating the trait kind of value (like is scalar or is sorted etc).

Here I think we need a different approach. I would first create a MultiJoinNode and then create a graph from it and then replace (inline) it with JoinNodes.

findepi replied:
I don't see how this would work. You mean incremental construction of `MultiJoinNode`-s from `Join` and other `MultiJoinNode` with one rule? You would need to "swallow" filters and projections too (it's what supported by the old optimizer). And you would need to know when to stop and allow some second rule to decompose that into `Join`-s. It doesn't play with current `IterativeOptimizer` well, does it? Wouldn't you need two separate `IterativeOptimizer`-s that _cannot_ be combined into one?

kokosing replied:
Yes. Yes. Yes. But this approach does not fit into iterative approach either. Let's discuss it offline.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/PlanVisitor.java:
```diff
@@ -186,4 +188,9 @@ public R visitAssignUniqueId(AssignUniqueId node, C context)
         return visitPlan(node, context);
     }
+
+    public R visitGroupReference(GroupReference node, C context)
```
Why do we need this? My original intention was to not have to ever use visitors for GroupReference nodes and to eventually get rid of them altogether.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -1046,6 +1047,14 @@ public Void visitAssignUniqueId(AssignUniqueId node, Integer indent)
         }
 
         @Override
+        public Void visitGroupReference(GroupReference node, Integer indent)
```
Is this just for debugging? GroupReference nodes should not exists outside of IterativeOptimizer. Therefore, they should never make it to an EXPLAIN call.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java:
```diff
@@ -35,6 +35,7 @@
 import java.util.HashSet;
```
Add an explanation of what's being fixed to the commit message, or make the commit title more explicit: "Fix incorrect equality check in EliminateCrossJoins"

findepi replied:
I renamed the commit (otherwise no code changes except for final fixups)

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java:
```diff
@@ -14,6 +14,7 @@
 package com.facebook.presto.sql.planner.optimizations.joins;
```
Does this commit fix a bug due to the migrated code not being able to operate on GroupReferences? If so, it should be part of the migration commit, as it's inherently part of the work required.

findepi replied:
You mean "Preserve leaf GroupReferences in EliminateCrossJoins rule" commit? I had this separate because

- without this commit code is still correct. It didn't lookup whole plan all the way down. It just looked up one level too far.
- test in this commit requires changes to rule testing framework (tests need to use lookup, we need to be able to match `GroupReference`-s and print them)
- i hoped reviewing will be easier

Anyway, if all gets accepted, let's squash. I renamed the commit to fixup.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java:
```diff
@@ -135,6 +136,78 @@ public static boolean isOriginalOrder(List<Integer> joinOrder)
                 .collect(toImmutableList());
     }
 
+    public static PlanNode createJoinsReplacement(PlanNode node, JoinGraph graph, List<Integer> joinOrder, PlanNodeIdAllocator idAllocator)
```
`node` is only used for its output symbols. Pass those instead ("expectedOutputs").

martint replied:
Maybe call this method "buildJoinTree", which more closely captures what the method does.

findepi commented:
Comments applied, rebased on master because there were conflicts.

findepi commented:
rebased on master to resolve conflicts

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java:
```diff
@@ -0,0 +1,63 @@
+        return Optional.of(replacement);
+    }
+
+    // TODO if the approach laid in PR is generally approved, move old EliminateCrossJoins optimizer's reused methods here,
```
What about this TODO?

findepi commented:
Thanks for the review!

> Have all comments from @kokosing been addressed?

There were. I only missed some about static imports, added now (and squashed already).

@kokosing wasn't sure the design of this rule fits iterative optimizer perfectly, but this you can judge yourself :)

> Can you squash the relevant commits before I merge it?

Squashed.

> What about this TODO?

Added cleanup commit moving the static helpers from old optimizer to the rule class, along with their tests.

martint commented:
Merged, thanks!
